src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -350,39 +351,104 @@ private void WriteAsPaxSharedInternal(Span<byte> buffer, out long actualLength)
         {
             actualLength = GetTotalDataBytesToWrite();
 
-            int tmpChecksum = WritePosixName(buffer);
+            int tmpChecksum = WriteName(buffer);
             tmpChecksum += WriteCommonFields(buffer, actualLength, TarHelpers.GetCorrectTypeFlagForFormat(TarEntryFormat.Pax, _typeFlag));
             tmpChecksum += WritePosixMagicAndVersion(buffer);
             tmpChecksum += WritePosixAndGnuSharedFields(buffer);
 
             _checksum = WriteChecksum(tmpChecksum, buffer);
         }
 
-        // All formats save in the name byte array only the ASCII bytes that fit.
+        // Gnu and pax save in the name byte array only the UTF8 bytes that fit.
+        // V7 does not support more than 100 bytes so it throws.
         private int WriteName(Span<byte> buffer)
         {
-            ReadOnlySpan<char> src = _name.AsSpan(0, Math.Min(_name.Length, FieldLengths.Name));
-            Span<byte> dest = buffer.Slice(FieldLocations.Name, FieldLengths.Name);
-            int encoded = Encoding.ASCII.GetBytes(src, dest);
-            return Checksum(dest.Slice(0, encoded));
+            ReadOnlySpan<char> name = _name;
+            int utf16NameTruncatedLength = GetUtf16TruncatedTextLength(name, FieldLengths.Name);
+
+            if (_format is TarEntryFormat.V7 && name.Length != utf16NameTruncatedLength)
+            {
+                throw new ArgumentException(SR.TarEntryNameExceedsMaxLength, ArgNameEntry);
+            }
+
+            ReadOnlySpan<char> truncatedName = name.Slice(0, utf16NameTruncatedLength);
+            Span<byte> destination = buffer.Slice(FieldLocations.Name, FieldLengths.Name);
+            int encoded = Encoding.UTF8.GetBytes(truncatedName, destination);
+
+            return Checksum(destination.Slice(0, encoded));
         }
 
-        // Ustar and PAX save in the name byte array only the ASCII bytes that fit, and the rest of that string is saved in the prefix field.
-        private int WritePosixName(Span<byte> buffer)
+        // 'https://www.freebsd.org/cgi/man.cgi?tar(5)'
+        // If the pathname is too long to fit in the 100 bytes provided by the standard format,
+        // it can be split at any / character with the first portion going into the prefix field.
```

Comment:
This says it can be split at any / character, but the code below is searching for / and \ on Windows.  Is that ok because we subsequently normalize \ to /? Or has that normalization already been done?

Reply:
Normalization has not been done yet, in fact, I don't think any normalization occurs on `TarWriter.WriteEntry`. @carlossanlop?

```cs
Fact]
    public void qt()
    {
        string name = "foo\\bar\\baz";
        TarEntry entry = new UstarTarEntry(TarEntryType.RegularFile, name);

        MemoryStream ms = new();
        TarWriter writer = new(ms, leaveOpen: true);
        writer.WriteEntry(entry);
        writer.Dispose();

        ms.Position = 0;
        using TarReader reader = new(ms);
        entry = reader.GetNextEntry();

        Console.WriteLine(entry.Name); // prints "foo\bar\baz".
        reader.Dispose();
    }
```